functions/Set-DbaDbCompression.ps1

```diff
@@ -131,167 +131,134 @@ function Set-DbaDbCompression {
                                 }
-                                if ($obj.indexId -le 1) {
-                                    ##heaps and clustered indexes
-                                    Write-Message -Level Verbose -Message "Applying $($obj.CompressionTypeRecommendation) compression to $($obj.Database).$($obj.Schema).$($obj.TableName)"
-                                    $($server.Databases[$obj.Database].Tables[$obj.TableName, $obj.Schema].PhysicalPartitions | Where-Object { $_.PartitionNumber -eq $obj.Partition }).DataCompression = $($obj.CompressionTypeRecommendation)
-                                    $server.Databases[$obj.Database].Tables[$obj.TableName, $($obj.Schema)].Rebuild()
-                                    $obj.AlreadyProcessed = "True"
-                                } else {
-                                    ##nonclustered indexes
-                                    Write-Message -Level Verbose -Message "Applying $($obj.CompressionTypeRecommendation) compression to $($obj.Database).$($obj.Schema).$($obj.TableName).$($obj.IndexName)"
-                                    $($server.Databases[$obj.Database].Tables[$obj.TableName, $obj.Schema].Indexes[$obj.IndexName].PhysicalPartitions | Where-Object { $_.PartitionNumber -eq $obj.Partition }).DataCompression = $($obj.CompressionTypeRecommendation)
+                            } else {
+                                ##nonclustered indexes
+                                Write-Message -Level Verbose -Message "Applying $($obj.CompressionTypeRecommendation) compression to $($obj.Database).$($obj.Schema).$($obj.TableName).$($obj.IndexName)"
+                                try {
+                                    ($server.Databases[$obj.Database].Tables[$obj.TableName, $obj.Schema].Indexes[$obj.IndexName].PhysicalPartitions | Where-Object { $_.PartitionNumber -eq $obj.Partition }).DataCompression = $obj.CompressionTypeRecommendation
                                     $server.Databases[$obj.Database].Tables[$obj.TableName, $obj.Schema].Indexes[$obj.IndexName].Rebuild()
-                                    $obj.AlreadyProcessed = "True"
+                                } catch {
+                                    Stop-Function -Message "Compression failed for $instance - $db - table $($obj.Schema).$($obj.TableName) - index $($obj.IndexName) - partition $($obj.Partition)" -Target $db -ErrorRecord $_ -Continue
                                 }
-                                $obj
                             }
+                            $obj.AlreadyProcessed = "True"
+                            $obj
+                        }
+                        # TODO: Should we output those objects that could not be processed due to the max run time? All objects or only those in the foreach?
```

Comment:
In practice, Set commands should only output what they changed.

Reply:
I agree. What is `.AlreadyProcessed` for - if it is always true in the output?